lib/assert.js

```diff
@@ -149,109 +139,123 @@ function isArguments(tag) {
-  // a) The same prototypes.
-  // b) The same built-in type tags
-  if (strict) {
-    if (Object.getPrototypeOf(actual) !== Object.getPrototypeOf(expected)) {
+  if (actualTag !== expectedTag) {
+    return false;
+  }
+  if (Object.getPrototypeOf(actual) !== Object.getPrototypeOf(expected)) {
+    return false;
+  }
+  if (isObjectOrArrayTag(actualTag)) {
+    return;
+  }
+  if (util.isDate(actual)) {
+    if (actual.getTime() !== expected.getTime()) {
       return false;
     }
-
-    if (actualTag !== expectedTag) {
+  } else if (util.isRegExp(actual)) {
+    if (!areSimilarRegExps(actual, expected)) {
+      return false;
+    }
+  } else if (!isFloatTypedArrayTag(actualTag) && ArrayBuffer.isView(actual)) {
+    if (!areSimilarTypedArrays(actual, expected)) {
       return false;
     }
-  }
 
-  // Do fast checks for builtin types.
-  // If they don't match, they must not be equal.
-  // If they match, return true for non-strict comparison.
-  // For strict comparison we need to exam further.
+    // Buffer.compare returns true, so actual.length === expected.length
+    // if they both only contain numeric keys, we don't need to exam further
+    if (Object.keys(actual).length === actual.length &&
+        Object.keys(expected).length === expected.length) {
+      return true;
+    }
+  }
+}
 
-  // If both values are Date objects,
-  // check if the time underneath are equal first.
+function looseDeepEqual(actual, expected) {
+  if (actual === null || typeof actual !== 'object') {
+    if (expected === null || typeof expected !== 'object') {
+      // eslint-disable-next-line eqeqeq
+      return actual == expected;
+    }
+    return false;
+  }
+  if (expected === null || typeof expected !== 'object') {
+    return false;
+  }
   if (util.isDate(actual) && util.isDate(expected)) {
     if (actual.getTime() !== expected.getTime()) {
       return false;
-    } else if (!strict) {
-      return true;  // Skip further checks for non-strict comparison.
     }
+    return true;
   }
-
-  // If both values are RegExp, check if they have
-  // the same source and flags first
   if (util.isRegExp(actual) && util.isRegExp(expected)) {
     if (!areSimilarRegExps(actual, expected)) {
       return false;
-    } else if (!strict) {
-      return true;  // Skip further checks for non-strict comparison.
     }
+    return true;
   }
-
-  // Ensure reflexivity of deepEqual with `arguments` objects.
-  // See https://github.com/nodejs/node-v0.x-archive/pull/7178
-  if (isArguments(actualTag) !== isArguments(expectedTag)) {
+  const actualTag = objectToString(actual);
+  const expectedTag = objectToString(expected);
+  if (actualTag === expectedTag) {
+    if (!isFloatTypedArrayTag(actualTag) && !isObjectOrArrayTag(actualTag) &&
+      ArrayBuffer.isView(actual)) {
+      if (!areSimilarTypedArrays(actual, expected)) {
+        return false;
+      }
+      return true;
+    }
+  } else if (isArguments(actualTag) || isArguments(expectedTag)) {
```

Comment:
This is a special case for compatibility so the link to the original PR should be kept IMO